### PR TITLE
Review

### PR DESCRIPTION
**Pull request by j-adamczyk**

### CODE REVIEW

File: PoC.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+import numpy as np
+from enum import Enum
```
Comment (j-adamczyk) [inline PoC.py:2]: Import order according to PEP8:
1. Built-in modules
2. 3rd party libraries
3. Your own project imports

Each group of imports should be separated with a space. Inside each group, `import...` statements should be first, and `from ... import` statements should be later. Each subgroup should be sorted lexicographically. For example:
```
import json
import os
from os import path

import numpy as np
import pandas as pd
from sklearn.model_selection import GridSearchCV
```

File: PoC.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+class DataType(Enum):
+    BINARY_SYMMETRIC = 0
+    BINARY_ASYMMETRIC = 1
+    CATEGORICAL_NOMINAL = 2
+    RATIO_SCALE = 3
+
+
+class GowerMetric:
+    def __init__(self, data_type: np.array):
+        self.__data_type = data_type        # initialize with np.array of column data types
```
Comment (j-adamczyk) [inline PoC.py:14]: https://scikit-learn.org/stable/developers/develop.html - "every keyword argument accepted by __init__ should correspond to an attribute on the instance". If you have `data_type` argument, you *have to* have `data_type` attribute. Additionally, maybe change to `dtypes`, similar to Pandas?

File: PoC.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        self.__iqrs = []                    # iqr values in .__ratio_scale()
+        self.__h = []                       # h values in .__ratio_scale()
```
Comment (j-adamczyk) [inline PoC.py:16]: https://scikit-learn.org/stable/developers/develop.html#estimated-attributes - "Attributes that have been estimated from the data must always have a name ending with trailing underscore, for example the coefficients of some regression estimator would be stored in a coef_ attribute after fit has been called."

File: PoC.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        self.__data_fitted = False          # flag for running .fit()
+        self.__sigma_sum = 0                # counter for existing values in vector
+
```
Comment (j-adamczyk) [inline PoC.py:19]: https://scikit-learn.org/stable/developers/develop.html#universal-attributes - add `n_features_in_` attribute

File: PoC.py (hunk 3 above)
Comment (j-adamczyk) [reply PoC.py:16]: Also note that those attributes have to be created in `.fit()`, somewhat counterintuitively: https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html - "Perform is_fitted validation for estimator.

Checks if the estimator is fitted by verifying the presence of fitted attributes (ending with a trailing underscore)"

File: PoC.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    CATEGORICAL_NOMINAL = 2
+    RATIO_SCALE = 3
+
+
+class GowerMetric:
+    def __init__(self, data_type: np.array):
+        self.__data_type = data_type        # initialize with np.array of column data types
+        self.__iqrs = []                    # iqr values in .__ratio_scale()
+        self.__h = []                       # h values in .__ratio_scale()
+        self.__data_fitted = False          # flag for running .fit()
```
Comment (j-adamczyk) [inline PoC.py:17]: Do not use an explicit attribute, as we can detect this with `sklearn.utils.validation.check_is_fitted()`, detecting presence of `self.iqrs_` and `self.h_` attributes

File: PoC.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        self.__sigma_sum = 0                # counter for existing values in vector
+
+    def __bin_sym(self, value_1, value_2) -> float:
```
Comment (j-adamczyk) [inline PoC.py:20]: Do not use double underscores, they are intended for avoiding inheritance conflicts. If you want a private method, usce a single underscore

File: PoC.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        if value_1 is None or value_2 is None:
+            return 0.0
```
Comment (j-adamczyk) [inline PoC.py:22]: Maybe check this in outer score, and only go inside a function if we know that each value exists? You have a lot of copied code currently

File: PoC.py (hunk 6 above)
Comment (j-adamczyk) [reply PoC.py:20]: Also, move private methods after public ones. This way to can quickly read through public API of a class, and helper functions, which are implementational details and are typically less interesting, are at the end.

File: PoC.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    RATIO_SCALE = 3
+
+
+class GowerMetric:
+    def __init__(self, data_type: np.array):
+        self.__data_type = data_type        # initialize with np.array of column data types
+        self.__iqrs = []                    # iqr values in .__ratio_scale()
+        self.__h = []                       # h values in .__ratio_scale()
+        self.__data_fitted = False          # flag for running .fit()
+        self.__sigma_sum = 0                # counter for existing values in vector
```
Comment (j-adamczyk) [inline PoC.py:18]: Why do we need this?

File: PoC.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                q1, q3 = np.percentile(x[:, i].astype(np.int32), [25, 75])
+                self.__iqrs.append(q3 - q1)
+
+                # h_t - bandwidth in the kernel density estimation (Marcello D’Orazio - p. 9)
+                if np.any(x[:, i].astype(np.int32) < 0):
+                    c = 0.9
+                else:
+                    c = 1.06
+
+                s = np.std(x[:, i].astype(np.int32))
```
Comment (j-adamczyk) [inline PoC.py:79]: 1. Why all this integer casting?
2. I would stick with `c=1.06` for now. Later we can add unimodality test to automatically detect `c` value

File: PoC.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                n = x[:, i].size
+                self.__h.append(c / n ** (1/5) * np.min([s, self.__iqrs[-1] / 1.34]))
+
+        self.__data_fitted = True
+        print("IQRS:", self.__iqrs)
+        print("h:", self.__h)
```
Comment (j-adamczyk) [inline PoC.py:85]: Remove prints, or add `verbose` option to constructor and print them if `verbose` is above 0

File: PoC.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    def __call__(
+        self, vector_1: np.array, vector_2: np.array,
+    ):
+        if len(vector_1) != len(vector_2):
```
Comment (j-adamczyk) [inline PoC.py:90]: Or if length of those vectors is different than `n_features_in`

File: PoC.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        return 1.0 - absolute / iqr
+
+    def __add_sigma(self):
+        self.__sigma_sum += 1
+
+    def __reset_sigma(self):
+        self.__sigma_sum = 0
+
+    def fit(self, x):
+        for i in range(self.__data_type.size):
```
Comment (j-adamczyk) [inline PoC.py:66]: 1. Set `n_features_in_`
2. You should use capital `X`

File: PoC.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+            if self.__data_type[i] == DataType.RATIO_SCALE:
+
```
Comment (j-adamczyk) [inline PoC.py:68]: Avoid empty lines after if, for etc.

File: PoC.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        print("h:", self.__h)
+
+    def __call__(
+        self, vector_1: np.array, vector_2: np.array,
+    ):
+        if len(vector_1) != len(vector_2):
+            print("Vector sizes don't match!")
+            return -1
+
+        if not self.__data_fitted:
```
Comment (j-adamczyk) [inline PoC.py:94]: Use `sklearn.utils.validation.check_is_fitted`. It will also automatically raise an appropriate exception if estimator is not fitted

File: PoC.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+                d += 1 - self.__bin_sym(vector_1[i], vector_2[i])
+            elif self.__data_type[i] == DataType.BINARY_ASYMMETRIC:
+                d += 1 - self.__bin_asym(vector_1[i], vector_2[i])
+            elif self.__data_type[i] == DataType.CATEGORICAL_NOMINAL:
+                d += 1 - self.__cat_nom(vector_1[i], vector_2[i])
+            elif self.__data_type[i] == DataType.RATIO_SCALE:
+                d += 1 - self.__ratio_scale(
+                    float(vector_1[i]), float(vector_2[i]), self.__iqrs[ratios_index], self.__h[ratios_index]
+                )
+                ratios_index += 1
```
Comment (j-adamczyk) [inline PoC.py:113]: 1. You can simplify this with a dictionary, mapping index in `self.dtypes` to function. This way you could do:
```
dist = lambda similarity, a, b: 1 - similarity(a, b)
distances = [dist(self.dtype_funcs[i], vector_1[i], vector_2[1]) for i in range(len(self.n_features_in_))]
distance = np.sum(distances)
```
2. Is `ratios_index` used anywhere?
3. I am pretty sure that we have to do more dtype casting here. Remember that Scikit-learn requires X array of dtype float32 as an input, so even for binary and categorical you won't have integers, but rather `0.` and `1.`. For those cases you have to cast appropriately, since you can't do direct comparison, i.e. `a == b` otherwise.

File: PoC.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+        self.__add_sigma()
+        absolute = np.abs(value_1 - value_2)
+
+        if absolute >= iqr:
+            return 0.0
+
+        elif absolute <= h:
+            return 1.0
+
+        return 1.0 - absolute / iqr
```
Comment (j-adamczyk) [inline PoC.py:57]: Remove empty lines, add `else`

File: PoC.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        if value_1 is None or value_2 is None:
+            return 0.0
+
+        if value_1 == value_2 and value_1 == 1:
+            self.__add_sigma()
+            return 1.0
+        else:
+            return 0.0
+
+    def __cat_nom(self, value_1, value_2) -> float:
```
Comment (j-adamczyk) [inline PoC.py:37]: This function and both binary functions require integers as input, otherwise you can't do `value_1 == value_2`, it won't work with floats.

Comment (j-adamczyk) [reply PoC.py:37]: Also, this can be vectorized. If you first take all elements that are categorical nominal, so you have two binary vectors `a` and `b`, you can do `a & b` to perform element-wise AND (so `value_1 == value_2`), and then sum: `(a & b).sum()`.

File: PoC.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+    def __bin_asym(self, value_1, value_2) -> float:
+        if value_1 is None or value_2 is None:
+            return 0.0
+
+        if value_1 == value_2 and value_1 == 1:
+            self.__add_sigma()
+            return 1.0
+        else:
+            return 0.0
```
Comment (j-adamczyk) [inline PoC.py:35]: This is just a Jaccard score, this can be vectorized and you can use Scipy function here: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.jaccard.html

File: PoC.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+    def __cat_nom(self, value_1, value_2) -> float:
+        if value_1 is None or value_2 is None:
+            return 0.0
+
+        self.__add_sigma()
+        return 1.0 if value_1 == value_2 else 0.0
+
+    def __ratio_scale(self, value_1, value_2, iqr, h) -> float:
```
Comment (j-adamczyk) [inline PoC.py:44]: We should support both traditional Gower distance normalization and IQR normalization. Maybe add `ratio_scale_normalization: str = "range"` to `__init__`, with an alternative value `"iqr"`?

File: PoC.py [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+        n = len(data_frame)
+        M = [[0.0 for _ in range(n)] for _ in range(n)]
+
+        for i in range(n):
+            for j in range(i):
+                if i != j:
+                    M[i][j] = self.__call__(data_frame[i], data_frame[j])
+                    M[j][i] = M[i][j]
+
+        return np.array(M, dtype=np.float32)
```
Comment (j-adamczyk) [inline PoC.py:133]: There already is a Scipy function for this: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist. This Gower metric should be usable with it, BTW.

File: PoC.py (hunk 17 above)
Comment (j-adamczyk) [reply PoC.py:37]: This can be merged with binary symmetric case, I think, since both really are a simple matching coefficient. This would mean that we can also vectorize them together.

File: pyproject.toml [hunk 21]
```diff
@@ -0,0 +1,16 @@
+[tool.poetry]
+name = "gowermetric"
+version = "0.1.0"
+description = ""
+authors = ["Arch4ngel21 <adammichal657@gmail.com>"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.9"
```
Comment (j-adamczyk) [inline pyproject.toml:9]: Add empty line between Python and further dependencies

File: pyproject.toml [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+numpy = "^1.23.3"
+black = "^22.8.0"
```
Comment (j-adamczyk) [inline pyproject.toml:11]: 1. Move black to dev dependencies: https://python-poetry.org/docs/master/managing-dependencies/#optional-groups
2. I personally prefer wildcard syntax, e.g. `numpy = "1.23.*"`. I think it is more explicit and similar to `requirements.txt`
3. Add pre-commit hook (https://pre-commit.com/) with black (https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/) to never forget about code formatting
